3d Tiles Inspector

austinEng opened the pull request:
Progress toward #4846. I would love to have some initial feedback on this

This replaces the debug options in the Sandcastle example http://localhost:8080/Apps/Sandcastle/index.html?src=3D%20Tiles.html&label=Showcases with an inspector widget. All previous functionality has been ported with the exception of the styling options.

Additional Todos:
- [x] Show URL of clicked tile
- [x] Click on tileset to inspect
- [x] Styling integration
- [x] Port other Sandcastle examples

Example initialization:
```javascript
viewer.extend(Cesium.viewerCesium3DTilesInspectorMixin);
var model = viewer.cesium3DTilesInspector.viewModel;

model.tileset = new Cesium.Cesium3DTileset({
    url: ...
});
```

Preview:
<img width="1433" alt="inspector" src="https://cloud.githubusercontent.com/assets/2154796/22025793/cd83d680-dc9c-11e6-92c4-ea0dd34b9aae.png">

pjcozzi commented:
> Styling integration?

Perhaps we could do a **super simple** style editor that shows the current style source and allows users to change it with automatic recompile.

pjcozzi commented:
@lilleyse can you please review the code?  I am only going to look at the functionality.  I'll do a quick glance at the graphics-related code at the end.

pjcozzi commented:
The Inspector should not include the dropdown of tilesets - that is specific to the Sandcastle example or someone's app.

Instead, leave the dropdown as part of the Sandcastle example and programmatic change the Inspector's focus when a new tileset is loaded.  We also want the ability for someone to pick on a tileset and open the Inspector (this might just be example code to do a pick then create the Inspector).

pjcozzi commented:
Perhaps remove "Picking" when "Enable Picking" is not checked.

![image](https://cloud.githubusercontent.com/assets/782098/22027723/b6d8bb52-dca2-11e6-9c80-54add82b742b.png)

pjcozzi commented:
> Perhaps remove "Picking" when "Enable Picking" is not checked.

Really, these are Sandcastle-specific features that I just added at some point for testing.  I don't think they need to be part of the inspector at all.  Will they help with debugging and profiling?  Maybe hide feature would, but I don't know that it is important enough.

These are probably better served as part of the Sandcastle example, perhaps just cleaned up compared to the original code.

pjcozzi commented:
@lilleyse when the "Request Volume" tileset is selected and "Request Volume" is checked, the volume disappears when the tile disappears.  Is this because of the SSE?  Or because the request volume failed?  If it's the later, should we still draw the request volume but perhaps another color?

![bv](https://cloud.githubusercontent.com/assets/782098/22029462/a8aa0972-dca8-11e6-8558-95f73a06fb8f.gif)

pjcozzi commented:
Include a text box so we can explicitly set the SSE.

The slider range should probably be wider, maybe 0 to 128.  Just a heads up that zero will load the entire tileset.

pjcozzi commented:
Can we make the stats easier to read?  Perhaps one per line with bullets and make the text (not numbers) bold?  Maybe look at Renderdoc for inspiration.

Also checking both "Stats" and "Pick Stats" hides the details of the "[Color]" pass:

![image](https://cloud.githubusercontent.com/assets/782098/22029758/843ec748-dca9-11e6-8167-fe64cfd021d2.png)

pjcozzi commented:
Great start, just those comments.

lilleyse commented:
> @lilleyse when the "Request Volume" tileset is selected and "Request Volume" is checked, the volume disappears when the tile disappears. Is this because of the SSE? Or because the request volume failed? If it's the later, should we still draw the request volume but perhaps another color?

It's the latter in this case - though the former could be true too. If the tile is not reached for any reason in `selectTiles` it won't add its debug command.

pjcozzi commented:
> It's the latter in this case - though the former could be true too. If the tile is not reached for any reason in selectTiles it won't add its debug command.

Do you think it is worth changing so that the debug volume can be shown?  I don't want to complicate the render loop or confuse the debug view (which is why I said to change the color).

lilleyse commented:
> Do you think it is worth changing so that the debug volume can be shown? I don't want to complicate the render loop or confuse the debug view (which is why I said to change the color)

It could be doable by keeping track of which tiles have request volumes. I'll think about how to cleanly add this.

pjcozzi commented:
>  I'll think about how to cleanly add this.

OK, but it is fine to bail if the tile selection code gets messy.

austinEng commented:
Okay, so I've updated and improved all of these things. Style editing is just a simple textarea that pops up. It reapplies the styles whenever the textarea loses focus. I figured that doing so on every keystroke would be too unstable unless there's a good way to verify that the styles are valid.

Tileset selection also works with `scene.pick` or the tileset can be manually set with `model.tileset = `

I'm not sure what other Sandcastle examples need to be modified, but I did add the inspector Batch Table Hierarchy and 3D Point Cloud Styling

pjcozzi commented:
> I'm not sure what other Sandcastle examples need to be modified, but I did add the inspector Batch Table Hierarchy and 3D Point Cloud Styling

Just those.

> Style editing is just a simple textarea that pops up. It reapplies the styles whenever the textarea loses focus. I figured that doing so on every keystroke would be too unstable unless there's a good way to verify that the styles are valid.

Perhaps add a "Compile" button that is highlighted whenever the text changes just like Sandcastle highlights the "Run" button.  Maybe also make a shortcut, "Compile (Ctrl-Enter)"

pjcozzi commented:
Please add unit tests to the [Specs/Widgets](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Specs/Widgets) directory.

pjcozzi commented:
Add a bullet to the "Added support for 3D Tiles..." part of [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/3d-tiles/CHANGES.md) since this is a public-facing change.

pjcozzi commented:
Can changing the SSE update the tileset as the slider is dragged, not just on mouse up?

pjcozzi commented:
Can the style editor show up just under the Style button instead of the other side of the screen?

pjcozzi commented:
Can the style editor catch exceptions when the style is compile and then display them like an IDE would just under the code editor?

In Sandcastle, they go to the console, which is great, but many users will be doing in-app debugging with this inspector outside of Sandcastle.

pjcozzi commented:
> Ignore batch table

Let's actually punt on this; if we end up needing it again, we'll add it.

pjcozzi commented:
On mouse over, do these also come up black in the `3d-tiles` branch?

![bv](https://cloud.githubusercontent.com/assets/782098/22124558/62e1c1ac-de5e-11e6-8239-2c7e0b00b89d.gif)

pjcozzi commented:
In the Batch Table Hierarchy example, this code does not appear to work, did it ever?  I don't think it is needed for this example, just the main 3D Tiles example.

```javascript
// When a feature is double middle clicked, hide it
handler.setInputAction(function(movement) {
    if (!model.picking) {
        return;
    }

    if (Cesium.defined(model.feature)) {
        model.feature.show = false;
    }
}, Cesium.ScreenSpaceEventType.MIDDLE_DOUBLE_CLICK);
```

pjcozzi commented:
In the Point Cloud Styling example, open the style editor and change the style, it doesn't seem to apply until you hit the "Close" button.

pjcozzi commented on Apps/Sandcastle/gallery/3D Tiles.html:
```diff
@@ -330,329 +272,6 @@
-    }
-
-    var menu = Sandcastle.addToolbarMenu([{
-        text : 'Color Blend Mode'
```
@lilleyse do you want to keep this for testing?  Or make it part of the inspector?

lilleyse replied:
It would be nice as part of the inspector.

pjcozzi replied:
Is there any other debug/test code in this file that should stay here or go into the inspector?

lilleyse replied:
This is probably good for now. One suggestion I have is for the inspector to return the tileset's properties so it is easy to know what variables to use to create styles.

lilleyse commented:
For whatever reason the double middle mouse event is just not triggering when I'm testing it, but the show/hide should still function.

I'm seeing the black transparency on `3d-tiles` too, I'll investigate.

pjcozzi commented:
Great progress, @austinEng!

Just those comments.  Note that I did not look at the inspector implementation.

lilleyse commented:
Left a comment in https://github.com/AnalyticalGraphicsInc/cesium/issues/2631 about the middle double click. I know this was working at least a few months ago, but I suppose we can't rely on it anymore.

pjcozzi commented:
@austinEng any update here?

Can you also please merge in master, there is a small conflict in the 3D Tiles Sandcastle example.

austinEng commented:
I just need to write some more tests for the inspector. Other than that I've updated based on all of your comments.

pjcozzi commented:
> "Suspend Updates"

I would call this "Freeze frame" like the API (`.debugFreezeFrame`) and perhaps update the terrain inspector if you want.

pjcozzi commented:
Checking "Dynamic SSE" hides the "Maximum SSE" textbox and slider.  I don't think it should.

pjcozzi commented:
Show **Stats** then switch to another tileset; it reports that 0 commands are generated:

![image](https://cloud.githubusercontent.com/assets/782098/22253726/af99a85a-e220-11e6-9fca-01bcba678566.png)

pjcozzi commented:
In #4846, it also had " Show the url of a clicked tile."  Let's wait to see if this is needed.  I _think_ it will be useful to identify tiles with artifacts, etc. that require separate inspection.

pjcozzi commented:
Crtl-Enter shortcut for applying a style doesn't work in the 3D Tiles Point Cloud Styling example.

pjcozzi commented:
> Crtl-Enter shortcut for applying a style doesn't work in the 3D Tiles Point Cloud Styling example.

Doesn't work in the 3D Tiles Batch Table Hierarchy example either.

pjcozzi commented:
I love the error console when a style doesn't compile, but it is hard to read.  Perhaps white text is better?

![image](https://cloud.githubusercontent.com/assets/782098/22254056/8fec01be-e221-11e6-8f01-09210c283e13.png)

austinEng commented:
> Crtl-Enter shortcut for applying a style doesn't work in the 3D Tiles Point Cloud Styling example.

I also noticed that it's not showing the properties of the tiles either. Not certain why this is. Are properties located in places other than tileset.properties, or are there default properties that are automatically added?

pjcozzi commented:
> I also noticed that it's not showing the properties of the tiles either. Not certain why this is. Are properties located in places other than tileset.properties, or are there default properties that are automatically added?

@lilleyse can help you track this down.

pjcozzi commented:
That's all my comments.

@austinEng please let @lilleyse know when this is ready for a final review and merge; I did not look at the code.  Make sure to merge in `3d-tiles` and write tests.

lilleyse commented:
> I also noticed that it's not showing the properties of the tiles either. Not certain why this is. Are properties located in places other than tileset.properties, or are there default properties that are automatically added?

I think those tilesets just aren't listing their properties in tileset.json. Eventually when https://github.com/AnalyticalGraphicsInc/3d-tiles-tools/pull/43 is in all the tilesets will be updated correctly. You can also get the properties from the picked feature through `feature.getPropertyNames`, but I wouldn't bother with that right now. Default properties are not automatically added.

lilleyse commented:
I also haven't looked closely at the code yet, but I ran through the controls some more and had some suggestions:

* Since our tileset urls often exceed the spacing given it would look better if the urls set their overflow to hidden instead of overflow to the next line.
* How easy would it be to add a vertical scrollbar to the inspector - as its hard to navigate when all panels are open. If this is not worth the effort I'm ok with passing on this.
* The title of each header is sometimes hard to distinguish from the rest of the content when expanded.
* I would remove the Requests/Attempted/Total and tiles styled / featured styled stats from the Pick Stats section. They are mostly redundant.
* Ctrl + Enter button needs a closing )
* Horizontal scroll for style editor would be nice instead of overflow
* Style editor should accept empty string as a valid style, and just turn it into `{}`. It should start this way by default when the demo doesn't add a style since its always nice to see a blank slate.
* 3D Tiles demo should load the first tileset at start
* Maximum SSE slider - better to increment by integers only - the .5 are distracting
* SSE Density values are traditionally very close to 0, whereas this slide is linear 0-1. Could the slider follow a logarithmic scale somehow? Either that of the slider is limited to a smaller range but the textbox can accept any value between 0 and 1.
* `Pick Tileset` - if no tileset is picked should it just ignore that the pick happened at all?
    * If you haven't already it would be a good idea to test this with multiple tileset to make sure the inspector plays well with that.

pjcozzi commented:
> Maximum SSE slider - better to increment by integers only - the .5 are distracting

However, we still want to be able to enter fractions explicitly in the text box.

austinEng commented:
> In #4846, it also had " Show the url of a clicked tile." Let's wait to see if this is needed. I think it will be useful to identify tiles with artifacts, etc. that require separate inspection

Is the url of a tile different from the url of a tileset because it currently shows the URL of the current tileset. Do you want this removed for now?

lilleyse commented:
> Is the url of a tile different from the url of a tileset because it currently shows the URL of the current tileset. Do you want this removed for now?

It's good to have both. The url for the tiles are different and are saved in the _url property of each `*3DTileContent`.

pjcozzi commented:
> Is the url of a tile different from the url of a tileset because it currently shows the URL of the current tileset. Do you want this removed for now?

Each tile has a separate url.

I don't find the url for the tileset to be all that useful and it is bulky, so perhaps remove it.  It could be useful in scenarios when lots of tilesets are loaded, but I do not know if that will be common.

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspector.js:
```diff
@@ -0,0 +1,364 @@
+        tilesetPanel.contents.appendChild(makeCheckbox('picking', 'Enable Picking'));
+        knockout.applyBindings(viewModel, tilesetPanel.contents);
+
+
```
Remove empty line.

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspector.js:
```diff
@@ -0,0 +1,364 @@
+        displayPanel.contents.appendChild(makeCheckbox('showRequestVolumes', 'Request Volumes'));
+        knockout.applyBindings(viewModel, displayPanel.contents);
+
+
```
Remove empty line here too.

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspector.js:
```diff
@@ -0,0 +1,364 @@
+
+        var dynamicSSEContainer = document.createElement('div');
+        dynamicSSEContainer.setAttribute('data-bind', 'css: {"cesium-cesiumInspector-show" : dynamicSSE, "cesium-cesiumInspector-hide" : !dynamicSSE}');
+        dynamicSSEContainer.appendChild(makeExponentialRangeInput(viewModel, 'dynamicSSEDensity', 0, 1, 200, 6  , 'SSE Density'));
```
`200, 6  , 'SS` remove space after `6`

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspector.js:
```diff
@@ -0,0 +1,364 @@
+                                           'value: colorBlendMode');
+        stylePanel.contents.appendChild(blendDropdown);
+
+
```
Here too

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspector.js:
```diff
@@ -0,0 +1,364 @@
+    /**
+     * Sets the callback function to call on tileset load
+     *
+     * @param {Function} callback the callback
```
Start descriptions with a capital: `@param {Function} callback The callback`

austinEng replied:
Actually removing these as they are never used. It doesn't make sense to register callbacks with the inspector when you can just register them on the tileset itself.

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspector.js:
```diff
@@ -0,0 +1,364 @@
+        }
+        button.setAttribute('data-bind', binding);
+
+
```
Remove empty line.

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspectorViewModel.js:
```diff
@@ -0,0 +1,701 @@
+             */
+            ignoreBatchTable : {
+                default: false
+            },
```
Are `showTileURL` and `ignoreBatchTable` going to be included in this PR? If not remember to remove them.

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspectorViewModel.js:
```diff
@@ -0,0 +1,701 @@
+                    if (defined(tileset)) {
+                        // that._statsLogger = that._updateStats.bind(that, false, false);
+                        // tileset.loadProgress.addEventListener(that._statsLogger);
+                        // tileset.allTilesLoaded.addEventListener(that._statsLogger);
```
Is this code needed anymore?

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspectorViewModel.js:
```diff
@@ -0,0 +1,701 @@
+                        that.colorBlendMode = tileset.colorBlendMode;
+
+                        that._updateStats(false, true);
+                        that._updateStats(true, true);
```
`_updateStats` only takes one argument (a few other areas do this too).

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspectorViewModel.js:
```diff
@@ -0,0 +1,701 @@
+                        if (current.feature !== feature) {
+                            if (defined(current.feature) &&
+                                defined(current.feature._batchTable) &&
+                                !current.feature._batchTable.isDestroyed()) {
```
Is the destroyed check needed anymore? Is this for loading without batch table?

austinEng replied:
when changing tilesets, the previous selected feature's batch table gets destroyed. Since it no longer exists to restore its previous color, we need to do this check.

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspectorViewModel.js:
```diff
@@ -0,0 +1,701 @@
+            get: function() {
+                var that = this;
+                return createCommand(function(sender, event) {
+                    if (event.ctrlKey && event.key === "Enter") {
```
Use single quotes for "Enter".

lilleyse replied:
Ctrl+Enter doesn't work for me still, event.key is "↵". Should this be `event.code`?

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspectorViewModel.js:
```diff
@@ -0,0 +1,701 @@
+        if (outputStats) {
+            // Since the pick pass uses a smaller frustum around the pixel of interest,
+            // the stats will be different than the normal render pass.
+            // var s = isPick ? '<strong>[Pick ]: </strong>' : '<strong>[Color]: </strong>';
```
Can this commented code be removed?

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspectorViewModel.js:
```diff
@@ -0,0 +1,701 @@
+        if (defined(this._tileset) && (!defined(this._style) || this._style !== this._tileset.style)) {
+            this._style = this._tileset.style;
+            if (defined(this._style)) {
+                this.styleString = JSON.stringify(this._style.style, null, '    ');
```
Two spaces is a lot better for readability with the small window.

lilleyse commented on Specs/Widgets/CesiumInspector/Cesium3DTilesInspectorSpec.js:
```diff
@@ -0,0 +1,117 @@
+    Globe,
+    createScene,
+    when) {
+    'use strict';
```
Fix up indentation for the includes.

lilleyse commented on Specs/Widgets/CesiumInspector/Cesium3DTilesInspectorSpec.js:
```diff
@@ -0,0 +1,117 @@
+    'Core/Ellipsoid',
+    'Scene/Globe',
+    'Specs/createScene',
+    'ThirdParty/when'
```
Besides the first item, list them alphabetically.

lilleyse commented on Specs/Widgets/CesiumInspector/Cesium3DTilesInspectorSpec.js:
```diff
@@ -0,0 +1,117 @@
+            expect(viewModel.statsText).not.toBe('');
+            viewModel.showStats = false;
+            expect(viewModel.statsText).toBe('');
+
```
Remove empty line.

lilleyse commented on Specs/Widgets/CesiumInspector/Cesium3DTilesInspectorViewModelSpec.js:
```diff
@@ -0,0 +1,227 @@
+    });
+
+    describe('display options', function() {
+
```
Remove empty line.

lilleyse commented on Specs/Widgets/CesiumInspector/Cesium3DTilesInspectorViewModelSpec.js:
```diff
@@ -0,0 +1,227 @@
+            return viewModel._tilesetLoaded;
+        });
+
+
```
Remove empty line.

biomassives commented:
This message was created automatically by mail delivery software.

A message that you sent could not be delivered to one or more of its
recipients. This is a temporary error. The following address(es) deferred:

  acmeideal@gmail.com
    Domain biomassiv.es has exceeded the max emails per hour (121/100 (121%)) allowed.  Message will be reattempted later

------- This is a copy of the message, including all the headers. ------
Received: from github-smtp2-ext1.iad.github.net ([192.30.252.192]:45116 helo=github-smtp2a-ext-cp1-prd.iad.github.net)
	by chi-server32.websitehostserver.net with esmtps (TLSv1.2:ECDHE-RSA-AES256-GCM-SHA384:256)
	(Exim 4.87)
	(envelope-from <noreply@github.com>)
	id 1cWoAY-003qcp-2E
	for greg@biomassiv.es; Thu, 26 Jan 2017 11:49:10 -0600
Date: Wed, 25 Jan 2017 06:49:48 -0800
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed; d=github.com;
	s=pf2014; t=1485355788;
	bh=/eXaTB6PO8s+EkX7CYlie76NtgF9pW6DJBbqBIjIxro=;
	h=From:Reply-To:To:Cc:In-Reply-To:References:Subject:List-ID:
	 List-Archive:List-Post:List-Unsubscribe:From;
	b=CxvQw0rM3f7aPHydTB53dNtGpNh4Gk9X5T2lSZs0tanPJI6f8ZGqpzfnNDVcsOAAl
	 SWs6M9dZH3Vxs3IvpycSGD6RTpXyEIE+jVdYv5l+UoevSqyH5Vctj3+XBez4SZWBVz
	 AyLDnjLZ8tJquE1uW6KFqU1dT7KzQluLjw8BsdEU=
From: Patrick Cozzi <notifications@github.com>
Reply-To: AnalyticalGraphicsInc/cesium <reply@reply.github.com>
To: AnalyticalGraphicsInc/cesium <cesium@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Message-ID: <AnalyticalGraphicsInc/cesium/pull/4872/c275127227@github.com>
In-Reply-To: <AnalyticalGraphicsInc/cesium/pull/4872@github.com>
References: <AnalyticalGraphicsInc/cesium/pull/4872@github.com>
Subject: Re: [AnalyticalGraphicsInc/cesium] 3d Tiles Inspector (#4872)
Mime-Version: 1.0
Content-Type: multipart/alternative;
 boundary="--==_mimepart_5888bb0c1fa2a_7e73ff53abbd1306843d";
 charset=UTF-8
Content-Transfer-Encoding: 7bit
Precedence: list
X-GitHub-Sender: pjcozzi
X-GitHub-Recipient: biomassives
X-GitHub-Reason: subscribed
List-ID: AnalyticalGraphicsInc/cesium
 <cesium.AnalyticalGraphicsInc.github.com>
List-Archive: https://github.com/AnalyticalGraphicsInc/cesium
List-Post: <mailto:reply@reply.github.com>
List-Unsubscribe: <mailto:unsub+0042d4e2f86e3007f4735fcfeceae9ed6e4ed165391e169f92cf0000000114a07d0c92a169ce0bffd8da@reply.github.com>,
 <https://github.com/notifications/unsubscribe/AELU4j2DoaXXciRcbszD2nhrPga4vZBZks5rV2EMgaJpZM4Llx6R>
X-Auto-Response-Suppress: All
X-GitHub-Recipient-Address: greg@biomassiv.es


----==_mimepart_5888bb0c1fa2a_7e73ff53abbd1306843d
Content-Type: text/plain;
 charset=UTF-8
Content-Transfer-Encoding: 7bit

> Is the url of a tile different from the url of a tileset because it currently shows the URL of the current tileset. Do you want this removed for now?

Each tile has a separate url.

I don't find the url for the tileset to be all that useful and it is bulky, so perhaps remove it.  It could be useful in scenarios when lots of tilesets are loaded, but I do not know if that will be common.

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/AnalyticalGraphicsInc/cesium/pull/4872#issuecomment-275127227
----==_mimepart_5888bb0c1fa2a_7e73ff53abbd1306843d
Content-Type: text/html;
 charset=UTF-8
Content-Transfer-Encoding: quoted-printable

<blockquote>
<p>Is the url of a tile different from the url of a tileset because it cu=
rrently shows the URL of the current tileset. Do you want this removed fo=
r now?</p>
</blockquote>
<p>Each tile has a separate url.</p>
<p>I don't find the url for the tileset to be all that useful and it is b=
ulky, so perhaps remove it.  It could be useful in scenarios when lots of=
 tilesets are loaded, but I do not know if that will be common.</p>

<p style=3D"font-size:small;-webkit-text-size-adjust:none;color:#666;">&m=
dash;<br />You are receiving this because you are subscribed to this thre=
ad.<br />Reply to this email directly, <a href=3D"https://github.com/Anal=
yticalGraphicsInc/cesium/pull/4872#issuecomment-275127227">view it on Git=
Hub</a>, or <a href=3D"https://github.com/notifications/unsubscribe-auth/=
AELU4v0_rEo2ZOkoXrzZW5MTOybD62jiks5rV2EMgaJpZM4Llx6R">mute the thread</a>=
.<img alt=3D"" height=3D"1" src=3D"https://github.com/notifications/beaco=
n/AELU4qFhhOlms6a_A_38MvxWBCgMMDajks5rV2EMgaJpZM4Llx6R.gif" width=3D"1" /=
></p>
<div itemscope itemtype=3D"http://schema.org/EmailMessage">
<div itemprop=3D"action" itemscope itemtype=3D"http://schema.org/ViewActi=
on">
  <link itemprop=3D"url" href=3D"https://github.com/AnalyticalGraphicsInc=
/cesium/pull/4872#issuecomment-275127227"></link>
  <meta itemprop=3D"name" content=3D"View Pull Request"></meta>
</div>
<meta itemprop=3D"description" content=3D"View this Pull Request on GitHu=
b"></meta>
</div>

<script type=3D"application/json" data-scope=3D"inboxmarkup">{"api_versio=
n":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name"=
:"GitHub"},"entity":{"external_key":"github/AnalyticalGraphicsInc/cesium"=
,"title":"AnalyticalGraphicsInc/cesium","subtitle":"GitHub repository","m=
ain_image_url":"https://cloud.githubusercontent.com/assets/143418/1749583=
9/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://c=
loud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed=
-b52498112777.png","action":{"name":"Open in GitHub","url":"https://githu=
b.com/AnalyticalGraphicsInc/cesium"}},"updates":{"snippets":[{"icon":"PER=
SON","message":"@pjcozzi in #4872: \u003e Is the url of a tile different =
from the url of a tileset because it currently shows the URL of the curre=
nt tileset. Do you want this removed for now?\r\n\r\nEach tile has a sepa=
rate url.\r\n\r\nI don't find the url for the tileset to be all that usef=
ul and it is bulky, so perhaps remove it.  It could be useful in scenario=
s when lots of tilesets are loaded, but I do not know if that will be com=
mon."}],"action":{"name":"View Pull Request","url":"https://github.com/An=
alyticalGraphicsInc/cesium/pull/4872#issuecomment-275127227"}}}</script>=

----==_mimepart_5888bb0c1fa2a_7e73ff53abbd1306843d--

austinEng commented:
@lilleyse updated

lilleyse commented:
> The annotation ability in the 3D Tiles demo is quite buggy right now when moving the camera, most likely not related to this PR but worth investigating.

Realize this is due to picking not supported for translucent features.

lilleyse commented on Source/Widgets/CesiumInspector/Cesium3DTilesInspectorViewModel.js:
```diff
@@ -0,0 +1,671 @@
+                        if (val) {
+                            that._eventHandler.setInputAction(function(e) {
+                                that._feature = scene.pick(e.endPosition);
+                                that._updateStats(true, false);
```
This one still passes 2 arguments.

austinEng commented:
> It would be nice for the style to be reapplied when clicking compile even if the text hasn't changed. Example: turn colorize on/off, then press compile style button - style should be reapplied even though it hasn't changed.

In addition to adding a force re-compile I think it would make sense to also reapply styles when colorize is disabled

pjcozzi commented:
@austinEng I just noticed 11d89a6c393bb61af93ade624d31ccf9ce2bfa57.  Is this ready?

austinEng commented:
@pjcozzi Not quite
@lilleyse Not certain how to forcefully reapply the styles after disabling debugColorizeTiles. Seems like marking the styleEngine as dirty doesn't work because the frameState is a render pass. I could probably loop though all the features and reapply the styles one-by-one, but that seems messy.

austinEng commented:
> Not certain how to forcefully reapply the styles after disabling debugColorizeTiles. Seems like marking the styleEngine as dirty doesn't work because the frameState is a render pass. I could probably loop though all the features and reapply the styles one-by-one, but that seems messy.

Okay, I figured this out. debug flags in the tiles aren't updated immediately upon changing flags of the tileset so things were getting reset.

I'm getting jsHint errors in `Specs/Renderer/TextureSpec.js` which I haven't modified. I'm guessing this happened after merging in 3d-tiles?

lilleyse commented:
> I'm getting jsHint errors in Specs/Renderer/TextureSpec.js which I haven't modified. I'm guessing this happened after merging in 3d-tiles?

Yeah don't worry about those. Once the texture compression PR is merged back into 3d-tiles they should be gone.

lilleyse commented:
It's probably worth addressing https://github.com/AnalyticalGraphicsInc/cesium/pull/4932#discussion_r98806592 before this PR is merged.

austinEng commented:
@lilleyse Ported the debug camera for freeze frame and fixed the on-click re-styling issue. The problem is that the Sandcastle example was setting the property 'clicked' to true/false which ends up marking all feature properties as dirty and causing a re-style.

I didn't really see the purpose of this 'clicked' property because all the example did was console.log if that feature had already been clicked so I removed it. Is this property needed elsewhere?

lilleyse commented:
The clicked setting is just example code for creating a property dynamically and then getting it. It isn't needed elsewhere. I don't really mind if we keep it or not.

pjcozzi commented:
What is the plan here?  Can we merge this before the bug bash on Wednesday?

austinEng commented:
I think this is finished, but the tests no longer pass after merging with 3d-tiles. However, I think @lilleyse said this was a problem elsewhere

lilleyse commented:
All looks good to me - @pjcozzi would you like review as well?

skyrs commented:
![image](https://cloud.githubusercontent.com/assets/25574370/22727150/46533566-ee12-11e6-8289-28d79d07c751.png)
I use 3d-tiles load my 3d model , but the result is very bad, so Who know why? I need set some property? my data is from osgb to b3dm , contain  Pagedlod

pjcozzi commented:
@skyrs Could be the lighting or materials.

Please ask questions like this on the [Cesium forum](http://cesiumjs.org/forum.html).  For more info on where to post, see [here](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#submitting-an-issue).
